rainier-core/src/main/scala/com/stripe/rainier/sampler/LeapFrog.scala

```diff
@@ -1,38 +1,88 @@
 package com.stripe.rainier.sampler
 
-import com.stripe.rainier.compute._
-import com.stripe.rainier.ir.CompiledFunction
-
-private[sampler] case class LeapFrog(nVars: Int,
-                                     initialHalfThenFullStep: CompiledFunction,
-                                     twoFullSteps: CompiledFunction,
-                                     finalHalfStep: CompiledFunction) {
+private[sampler] case class LeapFrog(density: DensityFunction) {
   /*
   Params layout:
   array(0..(n-1)) == ps
   array(n..(n*2-1)) == qs
   array(n*2) == potential
   array(n*2+1) == stepSize
    */
+  private val nVars = density.nVars
   private val potentialIndex = nVars * 2
   private val stepSizeIndex = potentialIndex + 1
   private val inputOutputSize = stepSizeIndex + 1
   private val buf1 = new Array[Double](inputOutputSize)
-  private val buf2 = new Array[Double](inputOutputSize)
 
-  private val globalsSize =
-    List(initialHalfThenFullStep, twoFullSteps, finalHalfStep)
-      .map(_.numGlobals)
-      .max
-  private val globals = new Array[Double](globalsSize)
+  def halfPs(inputs: Array[Double]): Unit = {
```

Comment:
these are always passed `buf1` so now that they're instance methods you don't need them to have a param

Reply:
Ha, `sbt` complained when I had nullary side-effecting methods, but these now take `stepSize` as a param so ✅